### ISSUE TITLE
Launchpad ppa deb build failiing on Ubuntu 14.04

### ISSUE THREAD

**Issue by andrewbird** [stsp/dosemu2#387]
The deb build is failing on Ubuntu 14.04. See the [buildlog](https://launchpadlibrarian.net/321457145/buildlog_ubuntu-trusty-i386.dosemu2_2.0~pre7-4462+844d536~ubuntu14.04.1_BUILDING.txt.gz) It fails to find keyboard.h and might be related to missing 'realpath' command.
~~~
gcc -std=gnu11 -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security" -I../../../../src/include -I../../../../src/plugin/include -I/<<PKGBUILDDIR>>/build/../src/include -D_FORTIFY_SOURCE=2  -o vesabios.o /<<PKGBUILDDIR>>/build/../src/base/dev/vga/vesabios.S
In file included from /<<PKGBUILDDIR>>/build/../src/include/emu.h:16:0,
                 from /<<PKGBUILDDIR>>/build/../src/base/dev/vga/miscemu.c:69:
/<<PKGBUILDDIR>>/build/../src/include/mouse.h:156:22: fatal error: keyboard.h: No such file or directory
 #include "keyboard.h"
                      ^
compilation terminated.
make[3]: *** [text.o] Error 1
make[3]: Leaving directory `/<<PKGBUILDDIR>>/build/src/base/video'
make[2]: *** [base/video] Error 2
make[2]: *** Waiting for unfinished jobs....
gcc -std=gnu11 -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security" -I../../../src/include -I../../../src/plugin/include -I/<<PKGBUILDDIR>>/build/../src/include -D_FORTIFY_SOURCE=2 -I. -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -o mkhdimage.o /<<PKGBUILDDIR>>/build/../src/tools/periph/mkhdimage.c
gcc -std=gnu11 -c -MMD -DCFLAGS_STR="-Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security" -I../../../../src/include -I../../../../src/plugin/include -I/<<PKGBUILDDIR>>/build/../src/include -D_FORTIFY_SOURCE=2 -Wall -Wstrict-prototypes -Wmissing-declarations -Wnested-externs -fms-extensions -pthread -imacros config.h -Wno-unused-result -fpie -fplan9-extensions -Wno-maybe-uninitialized -ggdb3   -g -O2 -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -o vgaemu.o /<<PKGBUILDDIR>>/build/../src/base/dev/vga/vgaemu.c
make[3]: *** [miscemu.o] Error 1
make[3]: *** Waiting for unfinished jobs....
In file included from /<<PKGBUILDDIR>>/build/../src/include/emu.h:16:0,
                 from /<<PKGBUILDDIR>>/build/../src/base/dev/vga/vgaemu.c:264:
/<<PKGBUILDDIR>>/build/../src/include/mouse.h:156:22: fatal error: keyboard.h: No such file or directory
 #include "keyboard.h"
                      ^
compilation terminated.
gcc -std=gnu11 -Wl,-warn-common -pthread -pie -Wl,-Bsymbolic-functions -Wl,-z,relro hdinfo.o -o hdinfo
gcc -std=gnu11 -Wl,-warn-common -pthread -pie -Wl,-Bsymbolic-functions -Wl,-z,relro mkhdimage.o -o ../../../2.0-pre6/bin/mkhdimage
gcc -std=gnu11 -Wl,-warn-common -pthread -pie -Wl,-Bsymbolic-functions -Wl,-z,relro mkfatimage16.o bootsect.o bootnorm.o -o ../../../2.0-pre6/bin/mkfatimage16
make[3]: Leaving directory `/<<PKGBUILDDIR>>/build/src/tools/periph'
make[3]: *** [vgaemu.o] Error 1
make[3]: Leaving directory `/<<PKGBUILDDIR>>/build/src/base/dev/vga'
make[2]: *** [base/dev/vga] Error 2
make[2]: Leaving directory `/<<PKGBUILDDIR>>/build/src'
make[1]: *** [default] Error 2
make[1]: Leaving directory `/<<PKGBUILDDIR>>/build'
dh_auto_build: make -j4 returned exit code 2
make: *** [build] Error 2
dpkg-buildpackage: error: debian/rules build gave error exit status 2
--------------------------------------------------------------------------------
Build finished at 20170527-2134

Finished
--------

E: Build failure (dpkg-buildpackage died)

~~~

earlier in the log there are whinges about realpath
~~~
dpkg-buildpackage
-----------------

dpkg-buildpackage: source package dosemu2
dpkg-buildpackage: source version 2.0~pre7-4462+844d536~ubuntu14.04.1
dpkg-buildpackage: source distribution trusty
 dpkg-source --before-build dosemu2-2.0~pre7
dpkg-buildpackage: host architecture i386
 fakeroot debian/rules clean
dh clean --parallel --builddirectory=build
   dh_testdir -O--parallel -O--builddirectory=build
   dh_auto_clean -O--parallel -O--builddirectory=build
   dh_clean -O--parallel -O--builddirectory=build
 debian/rules build
dh build --parallel --builddirectory=build
   dh_testdir -O--parallel -O--builddirectory=build
   debian/rules override_dh_auto_configure
make[1]: Entering directory `/<<PKGBUILDDIR>>'
./autogen.sh
Generating toplevel configure script...
./mkpluginhooks: 1: ./mkpluginhooks: realpath: not found
cat: plugin_configure: No such file or directory

Done, now run ./default-configure
~~~

On other Ubuntu versions(e.g. 16.04 here) that have realpath as part of the coreutils package, the build completes fine
~~~
$ dpkg -S /usr/bin/realpath
coreutils: /usr/bin/realpath
~~~

On 14.04 realpath is seemingly packaged separately [here](https://packages.ubuntu.com/trusty/realpath)

**Comment by stsp** [comment]
If I depend on realpath, will it not break
other builds?

**Comment by andrewbird** [comment]
I don't think so, I just installed 'realpath' package on my 16.04 system and it seems to only contain docs now
~~~
root# apt install realpath
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following NEW packages will be installed
  realpath
0 to upgrade, 1 to newly install, 0 to remove and 0 not to upgrade.
Need to get 1,088 B of archives.
After this operation, 24.6 kB of additional disk space will be used.
Get:1 http://gb.archive.ubuntu.com/ubuntu xenial/main i386 realpath all 8.25-2ubuntu2 [1,088 B]
Fetched 1,088 B in 1s (978 B/s)      
Selecting previously unselected package realpath.
(Reading database ... 466541 files and directories currently installed.)
Preparing to unpack .../realpath_8.25-2ubuntu2_all.deb ...
Unpacking realpath (8.25-2ubuntu2) ...
Setting up realpath (8.25-2ubuntu2) ...
root# which realpath
/usr/bin/realpath
root# dpkg -S /usr/bin/realpath
coreutils: /usr/bin/realpath
root# dpkg -l realpath
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-=================================
ii  realpath       8.25-2ubuntu all          coreutils realpath transitional p
root# dpkg -L realpath
/.
/usr
/usr/share
/usr/share/doc
/usr/share/doc/realpath
/usr/share/doc/realpath/copyright
/usr/share/doc/realpath/changelog.Debian.gz
~~~

**Comment by stsp** [comment]
Should now be fixed then.

**Comment by andrewbird** [comment]
Thanks, all versions are building successfully now.

**Comment by stsp** [comment]
I wonder why everything suddenly broke.
I didn't do anything for a few weeks...
Perhaps software ages when you don't
touch it.

**Comment by andrewbird** [comment]
I think the first failure messages started appearing on the 22nd, but I didn't get around to reporting it for a while :-( . Perhaps I could add you to the dosemu2 group on the Launchpad, then you'd get the messages as soon as a build has issues?

**Comment by stsp** [comment]
Why not? Please do, thanks.
I also now fixed a few more problems that
I've spotted in the build log.

**Comment by stsp** [comment]
I wonder what this means:
```
dh_strip debug symbol extraction: disabling for PPA build
dh_strip debug symbol extraction: not enabling -a because /CurrentlyBuilding exists
dh_strip debug symbol extraction: not doing anything since NO_PKG_MANGLE is given
```

But at the end it still provides an info about
a debuginfo package, as if it was built.

**Comment by andrewbird** [comment]
I added you to the dosemu2 team on Launchpad, feel free to change what you wish.

Sorry I'm not familiar with building separate debug symbol packages, so I have no idea.

**Comment by stsp** [comment]
OK, thanks!

**Comment by andrewbird** [comment]
@stsp it seems the launchpad import facility was having problems converting the git repo to bazaar and was failing to update the ppa for the last month. I've deleted the bazaar based repo/recipe and made new git based ones. It seems to be working okay so far, but it you see any problems it's likely related to this switch.

**Comment by stsp** [comment]
That's strange, I was following the notifications
and fixed all the problems. Could you please show
the logs, or are they already lost?
Is it possible for the new repo to use something
like `git clone -b devel --single-branch`
to not take all branches?

**Comment by andrewbird** [comment]
The bazaar import was failing rather than the build,.I think the logs are gone, but it was something to do with a git reference it didn't understand. I thought it would be easier to have a native git repo to avoid this type of issue.  I did ask launchpad web tool to only import the devel branch, but all branches appeared anyway. Feel free to have a poke around to see if you can create a single branch import, it's easy enough to delete and start over if necessary.

**Comment by stsp** [comment]
```
stsp commented on 28 May 2017
If I depend on realpath, will it not break
other builds?
@andrewbird
andrewbird commented on 28 May 2017
I don't think so, I just installed 'realpath' package on my 16.04 system and it seems to only contain docs now
```

Andrew, that appeared wrong: it broke bionic, so
I removed realpath again in 418c1e100, Tue Mar 13. 2018.
But Trusty was already broken with 15b806e8
at Tue Oct 10, 2017 as its gcc doesn't support
force_align_arg_pointer attr. Commit a552dd0e
added the configure check for that at Fri May 4, 2018.
Since realpath dep was removed month before,
it didn't work. Today I got rid of realpath dep
completely. And what? Nothing! fdpp was added
a week ago, so Trusty fails again.
This looks like a wild goose chase. No matter
what I fix, Trusty finds its ways to break.
I think I should stop fixing trusty forever.
